besu/src/test/java/org/hyperledger/besu/PrivacyReorgTest.java

```diff
@@ -0,0 +1,558 @@
+            privateStateStorage.getPrivacyGroupHeadBlockMap(blockchain.getGenesisBlock().getHash()))
+        .isEmpty();
+    assertThat(privateStateStorage.getPrivacyGroupHeadBlockMap(firstBlock.getHash())).isNotEmpty();
+    assertThat(privateStateStorage.getPrivacyGroupHeadBlockMap(firstBlock.getHash()))
+        .contains(expected);
+
+    final Block secondBlock =
+        gen.block(
+            new BlockDataGenerator.BlockOptions()
+                .setBlockNumber(2)
+                .setParentHash(firstBlock.getHeader().getHash())
+                .hasTransactions(false)
+                .hasOmmers(false)
+                .setReceiptsRoot(
+                    Hash.fromHexString(
+                        "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421"))
+                .setGasUsed(0)
+                .setLogsBloom(LogsBloomFilter.empty())
+                .setStateRoot(
+                    Hash.fromHexString(
+                        "0xd86a520e49caf215e7e4028262924db50540a5b26e415ab7c944e46a0c01d704")));
+
+    appendBlock(besuController, blockchain, protocolContext, secondBlock);
+
+    assertThat(privateStateStorage.getPrivacyGroupHeadBlockMap(secondBlock.getHash())).isNotEmpty();
+
+    assertThat(privateStateStorage.getPrivacyGroupHeadBlockMap(firstBlock.getHash()))
+        .contains(expected);
+  }
+
+  @Test
+  public void reorgToChainAtEqualHeight() {
+    // Setup an initial blockchain with one private transaction
+    final ProtocolContext<?> protocolContext = besuController.getProtocolContext();
+    final DefaultBlockchain blockchain = (DefaultBlockchain) protocolContext.getBlockchain();
+
+    final Block firstBlock =
+        gen.block(
+            new BlockDataGenerator.BlockOptions()
+                .setBlockNumber(1)
+                .setParentHash(blockchain.getGenesisBlock().getHash())
+                .addTransaction(generateMarker(getEnclaveKey(enclave.clientUrl())))
+                .hasOmmers(false)
+                .setReceiptsRoot(
+                    Hash.fromHexString(
+                        "0xc8267b3f9ed36df3ff8adb51a6d030716f23eeb50270e7fce8d9822ffa7f0461"))
+                .setGasUsed(23176)
+                .setLogsBloom(LogsBloomFilter.empty())
+                .setStateRoot(
+                    Hash.fromHexString(
+                        "0x16979b290f429e06d86a43584c7d8689d4292ade9a602e5c78e2867c6ebd904e")));
+
+    appendBlock(besuController, blockchain, protocolContext, firstBlock);
+
+    // Check that the private state root is not the empty state
+    assertPrivateStateRoot(
+        privateStateRootResolver,
+        blockchain,
+        "0x2121b68f1333e93bae8cd717a3ca68c9d7e7003f6b288c36dfc59b0f87be9590");
+
+    // Create parallel fork of length 1 which removes privacy marker transaction
+    final BlockDataGenerator.BlockOptions options =
+        new BlockDataGenerator.BlockOptions()
+            .setBlockNumber(firstBlock.getHeader().getNumber())
```

Comment:
This will always be 1, right? Using the number will make it a bit easier to read.